core/internal/filetransfer/task_default.go

```diff
@@ -0,0 +1,99 @@
+package filetransfer
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+)
+
+type DefaultTask struct {
+	// FileKind is the category of file being uploaded or downloaded
+	FileKind RunFileKind
+
+	// Path is the local path to the file
+	Path string
+
+	// Name is the name of the file
+	Name string
+
+	// Url is the endpoint to upload to/download from
+	Url string
+
+	// Headers to send on the upload
+	Headers []string
+
+	// Size is the number of bytes to upload
+	//
+	// If this is zero, then all bytes starting at `Offset` are uploaded; if non-zero,
+	// then that many bytes starting from `Offset` are uploaded.
+	Size int64
+
+	// Offset is the beginning of the file segment to upload
+	Offset int64
+
+	// Response is the http.Response from a successful upload or download request.
+	//
+	// This is nil for failed requests, or requests that have not completed.
+	Response *http.Response
+
+	// Error, if any.
+	Err error
+
+	// Callback to execute after completion (success or failure).
+	CompletionCallback func(Task)
+
+	// ProgressCallback is a callback to execute on progress updates
+	ProgressCallback func(int, int)
+
+	// This can be used to cancel the file upload or download if it is no longer needed.
+	Context context.Context
+}
+
+func (t *DefaultTask) GetFileKind() RunFileKind          { return t.FileKind }
+func (t *DefaultTask) GetPath() string                   { return t.Path }
+func (t *DefaultTask) GetUrl() string                    { return t.Url }
+func (t *DefaultTask) GetSize() int64                    { return t.Size }
+func (t *DefaultTask) GetErr() error                     { return t.Err }
+func (t *DefaultTask) GetCompletionCallback() func(Task) { return t.CompletionCallback }
+
+func (t *DefaultTask) SetProgressCallback(callback func(int, int)) {
+	t.ProgressCallback = callback
+}
+
+func (t *DefaultTask) SetCompletionCallback(callback func(Task)) {
+	t.CompletionCallback = callback
+}
+
+func (t *DefaultTask) SetErr(err error) {
+	t.Err = err
+}
+
+type DefaultUploadTask struct{ DefaultTask }
```

Comment:
I think `type DefaultUploadTask DefaultTask` works as well (note no `=` between the names, so that it's a new type rather than an alias)

Reply:
I tried this at first but only the fields were inherited, so `DefaultUploadTask` didn't implement `Task` unless I added the other methods again 🤔

Reply:
That's fine---Task shouldn't have so many methods anyway. I think it's possible to remove most of them; see my other comments. All you really need is `Execute` and `Complete`. If you pass `FileTransferStats` to `Complete`, then you don't even need to expose `FileKind` / `Path` / `Size` on the public interface.

Reply:
yeah that makes sense, I'll add `Complete` first and see how I can get rid of the other methods